### PR TITLE
ADDR-129: Refactor AH to allow loading configs via external modules

### PR DESCRIPTION
**Pull request by Ruhanga**
Ticket: https://issues.openmrs.org/browse/ADDR-129

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java
Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: Thanks for the review @mogoodrich. `initializeFullAddressCache()` depends on the loaded address config and so would initialize against non-existing address entries this is why I removed it. Rarely is the context refreshed without restarting an OpenMRS instance and therefore Iniz would just reload/refresh the changed cache at restart IMO.

File: api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java
Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:47]: `initI18nCache()` depends on the loaded address config as well and so would initialize against non-existing address entries which is why I removed it.

Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:47]: @mogoodrich you might also want to have a look at the other way this is being preferably consumed with this other [PR on Iniz](https://github.com/mekomsolutions/openmrs-module-initializer/pull/254)

File: api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: Thanks @Ruhanga !  Where does this happen in the Iniz PR?  This cache initialization needs to happen on every restart (not just went the address hierarchy csv changes or is iniitial loaded) so it seems odd that we'd have Iniz handle this?

File: api/src/main/java/org/openmrs/module/addresshierarchy/config/AddressConfigurationLoader.java
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/addresshierarchy/config/AddressConfigurationLoader.java:131]: Ah, nevemind, I realize now that you are calling this method from Iniz

File: api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java
Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: @mogoodrich I've added a comment [here](https://github.com/mekomsolutions/openmrs-module-initializer/pull/254#discussion_r1378431685) on the PR, where Iniz is invoking AH API.

Comment (mseaton) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: I agree with @mogoodrich and have added comments to the Iniz PR that reflect this same thing.  Basically, we need the 2 caches here to be initialized a) after the AH entries in the database have been updated and so the cache is stale, and b) any time the system is restarted and the cache has not yet been initialized.  I think if we add the initialization of these 2 caches into the AddressConfigurationLoader at the end to be invoked if either of these scenarios is true, then we can eliminate the calls here as you are doing, but also ensure it is still done when needed via having the AddressConfigurationLoader invoked from Initializer.

File: api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java
Comment (mseaton) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:47]: See comments above.  Same apply here

File: api/src/main/java/org/openmrs/module/addresshierarchy/config/AddressConfigurationLoader.java
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/addresshierarchy/config/AddressConfigurationLoader.java:52]: I think we can eliminate passing in the "domain" here.  That seems redundant and unnecessary.

**Review by mseaton — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: > Rarely is the context refreshed without restarting an OpenMRS instance and therefore Iniz would just reload/refresh the changed cache at restart IMO.

Modules should be designed to properly work without a restart. It's because we remove things like this that we force people to reboot when they shouldn't need to. The module reloading functionality isn't perfect, but it's unclear to me that removing this has _any_ advantage other than crippling the ability to reload the module without a restart.

Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: @ibacher, before these changes, there is these two things happening when AH is co-running with Iniz on the same instance. 1) AH will always load the `addresshierarchy` proxy domain config before Iniz can do any config loading.  2.) AH blindly depends on Iniz loading of message properties that it should consume for it's `initI18nCache()`. This results in somewhat a cyclic dependency that ends in an empty i18n cache. 

What I was trying to achieve with these changes is, to allow for a separation of concerns where all the metadata loading/cache initialization is all done with Iniz and AH providing the API to do, so we get more predictable results.

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: I get that. If we're talking about caching, though, AH should be able to do it, even if we expose an API so that Iniz (which will run afterwords) can re-run the caching, right? That stops Iniz from being a critical dependency for this module to work.

Basically, unless building an empty cache is an expensive procedure, I don't see that harm in running it twice and we don't end up in the bad state where AH _requires_ Iniz but cannot declare this dependency.

Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: One way of getting out of this cyclic dependency is checking for whether on not Iniz is loaded and consequentially delegating to it (Iniz) the loading of the AH related config (including cache initializations). This way we won't need an major version upgrade since the changes will be backwards compatible.

Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: From the wide discussion on the whole topic, I guess I'll have to simply add both cache initialization procedures at the end of `AddressConfigurationLoader.loadAddressConfiguration(...)` so they are always executed. This way Iniz will be able to implicitly invoke them upon every restart wether checksums have changed or not. The caches would then be initialized with the message properties that Iniz loads. 

@mogoodrich, @ibacher, @mseaton I think this would suffice and remain backwards compatible without introducing a brutal requirement on the Iniz module to load the AH configs?

Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: @Ruhanga  defer to you on the 18n cache, as I'm not familiar with how this works, but I still don't understand why keeping the line that initiailizes the fullAddressCache in the the contextRefreshed of the Address Hierarchy Activator introduced any kind of requirement on the Iniz module to load the AH config, but perhaps I'm missing something?

For the fullAddressCache, I see the following workflows:

**For a new install:**

1. Context refresh in AH Activator calls initializeFullAddressCache; since there are no entries in the AH tables, nothing happens (the initializeFullAddressCache checks to make sure there are entries in the database before doing anything)
2. Iniz calls the loadAddressConfiguration method, which loads the csv files, and then triggers a reload of the initializeFullAddressCache

**For a restart with no changes**

1. Context refresh in AH Activator calls initializeFullAddressCache; since there are entries in the AH tables, and the cache hasn't been initialize, the full address cache will be initialized
2. Iniz will call loadAddressConfiguration, and since there are no changes to the csv, it should do nothing (ie not reload address cache)

**For a restart with changes**

1. Context refresh in AH Activator calls initializeFullAddressCache; since there are entries in the AH tables, and the cache hasn't been initialize, the full address cache will be initialized
2. Iniz will call loadAddressConfiguration, and since there are changes to the csv, it will clear out the cache and the databaes, and load the new csvs, and then load the cache.

I guess there's a bit of inefficiency in this last case, because there's no need to load the cache in step 1 if it's going to be reloaded in step 2, but I'm fine with that, and would prefer this workflow rather than, in essentce, making the Address Hierarchy module dependent on Iniz to get up it's full address cache properly.

Comment (Ruhanga) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: Correct observations @mogoodrich. It's fine with me keeping the redundancy in invoking `initializeFullAddressCache()`. The problem was rather more to do with `initI18nCache()` that expects i18n message properties that are loaded by Iniz (unless those are available on the AH's classpath at compile time). 

For all the three workflows above, _message properties loading_ is done by Iniz and therefore, until Iniz has started, those properties are not available at the time AH wants to consume them inside `initI18nCache()`. One way to resolve this as implied from the previous discussions is to always invoke `initI18nCache()` at the end of `AddressConfigurationLoader.loadAddressConfiguration(...)`.

Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/addresshierarchy/AddressHierarchyActivator.java:46]: Great, thanks @Ruhanga ! I'll defer to you and others on the best way to take care of the I18nCache, but let's keep the fullAddressCache as it currently works, thanks!

**Review by mogoodrich — APPROVED** [review]
LGTM... thanks for addng the initializeFullAddressCache back in, I think it's better that way.

However, we *do* still want to remove the AddressConfigurationLoader.loadAddressConfiguration() call from the started method, correct?  It looks like that is no longer being removed?

**Comment by Ruhanga** [comment]
>we do still want to remove the AddressConfigurationLoader.loadAddressConfiguration() call from the started method, correct? 

Since we want the module to be usable without a hard requirement on Iniz, I'm happy leaving the method in. I've simply introduced a flavour of `loadAddressConfiguration(...)` that Iniz can interface with conveniently.

**Comment by mogoodrich** [comment]
Thanks @Ruhanga !  I'm happy to merge this in an test if @ibacher  and @mseaton  have no further comments.

**Comment by Ruhanga** [comment]
Hi @ibacher / @mseaton,  any objection getting this merged in?

**Review by mseaton — APPROVED** [review]
Looks fine to me.  It looks like the only real meaningful change is that you are now initializing the i18n cache whenever there is a call to load the address configuration.

**Comment by Ruhanga** [comment]
Correct @mseaton + allowing for other APIs like Iniz to provide the configuration/checksums directories.

**Review by ibacher — APPROVED** [review]
Thanks for this @Ruhanga!

**Pull request merged**
